hooks/post-command

```diff
@@ -33,90 +33,145 @@ display_success() {
+        digest=$(docker inspect --format='{{index .RepoDigests 0}}' "${image}")
+
+        status=$?
+        if [[ $status -ne 0 ]]; then
+            display_error "docker inspect" "Failed to get the local image digest, will continue using supplied image reference ${image}"
+        else
+            display_success "docker inspect" "Will continue using ${digest}"
+            image="${digest}"
+        fi
+    fi
+}
 
-# Parameters
-############
+# Common parameters
+###################
 
-# This is a required parameter
+# image is a required parameter
 image=${BUILDKITE_PLUGIN_COSIGN_IMAGE}
 if [[ -z "${image}" ]]; then
-    fail_with_message "cosign" "No image specified"
+    fail_with_message "cosign" "Image not specified"
 fi
+use_image_digest
+
+# flags for the cosign sign command
+sign_flags=("-y" "--output-signature" "out.sig")
 
 is_keyless=${BUILDKITE_PLUGIN_COSIGN_KEYLESS:-true}
 
 # Hook functions
 ################
 
-cosign_keyless() {
-    local fulcio_url=${BUILDKITE_PLUGIN_COSIGN_KEYLESS_CONFIG_FULCIO_URL:-"https://fulcio.sigstore.dev"}
-    local rekor_url=${BUILDKITE_PLUGIN_COSIG_KEYLESS_CONFIGN_REKOR_URL:-"https://rekor.sigstore.dev"}
-    local oidc_issuer=${BUILDKITE_PLUGIN_COSIG_KEYLESS_CONFIGN_OIDC_ISSUER:-"https://oauth2.sigstore.dev/auth"}
-    local oidc_provider=${BUILDKITE_PLUGIN_COSIG_KEYLESS_CONFIGN_OIDC_PROVIDER:-"buildkite-agent"}
-    
-    echo "--- :key: Cosign keyless signing"
-
-    rm out.sig || true
-
-    COSIGN_EXPERIMENTAL=1 cosign sign \
-        -y \
-        --fulcio-url="${fulcio_url}" \
-        --rekor-url="${rekor_url}" \
-        --oidc-issuer="${oidc_issuer}" \
-        --oidc-provider="${oidc_provider}" \
-        --output-signature=out.sig \
-        "${image}"
+# if provided, initialise cosign with a custom TUF configuration
+cosign_init() {
+    echo "--- :key: Init cosign"
 
-    status=$?
-    if [[ $status -ne 0 ]]; then
-        fail_with_message "cosign" "Failed to sign image"
+    # flags for the cosign initialize command
+    init_flags=()
+
+    if [[ "${is_keyless}" == true ]]; then
+        local tuf_mirror_url=${BUILDKITE_PLUGIN_COSIGN_KEYLESS_CONFIG_TUF_MIRROR_URL}
+        local tuf_root_url=${BUILDKITE_PLUGIN_COSIGN_KEYLESS_CONFIG_TUF_ROOT_URL}
+    else
+        local tuf_mirror_url=${BUILDKITE_PLUGIN_COSIGN_KEYED_CONFIG_TUF_MIRROR_URL}
+        local tuf_root_url=${BUILDKITE_PLUGIN_COSIGN_KEYED_CONFIG_TUF_ROOT_URL}
     fi
 
-    local signature=$(cat out.sig)
+    if [[ -n "${tuf_mirror_url}" ]]; then
+        init_flags+=("--mirror" "${tuf_mirror_url}")
+    fi
 
-    display_success "cosign" "Successfully signed image."
-    cat <<EOF | buildkite-agent annotate --style success --context "cosign-signature"
-### Signed image
-\`\`\`
-$image
-\`\`\`
+    if [[ -n "${tuf_root_url}" ]]; then
+        init_flags+=("--root" "${tuf_root_url}")
+    fi
 
-### Signature
-\`\`\`
-$signature
-\`\`\`
-EOF
+    if [ ${#init_flags[@]} -gt 0 ]; then
+        rm -rf ~/.sigstore
```

Comment:
I am wondering if we need to delete the whole directory. It would be unfortunate if there are other important things in there. The documentation for `cosign initialize` seems to claim that it updates the tuf thing:

> Any updated TUF repository will be written to $HOME/.sigstore/root/.
> https://github.com/sigstore/cosign/blob/main/doc/cosign_initialize.md

It looks like it gets overwritten:

```console
$ rm -rf ~/.sigstore/
$ cosign initialize --mirror https://foo.com --root https://foo.com/root.json
(...)
$ ls -l ~/.sigstore/root/
total 10
-rw-------  1 x x 56 Jun 18 21:31 remote.json
drwx------  2 x x  5 Jun 18 21:31 targets
drwxr-xr-x  2 x x  8 Jun 18 21:31 tuf.db
$ date
Tue Jun 18 21:31:55 UTC 2024
$ date
Tue Jun 18 21:32:02 UTC 2024
$ cosign initialize --mirror https://foo.com --root https://foo.com/root.json
(...)
$ ls -l ~/.sigstore/root/
total 10
-rw-------  1 x x 56 Jun 18 21:32 remote.json
drwx------  2 x x  5 Jun 18 21:31 targets
drwxr-xr-x  2 x x  9 Jun 18 21:32 tuf.db
$ ls -l ~/.sigstore/root/tuf.db/
total 16
-rw-r--r--  1 x x 2955 Jun 18 21:32 000004.ldb
-rw-r--r--  1 x x 4928 Jun 18 21:32 000007.log
-rw-r--r--  1 x x   16 Jun 18 21:32 CURRENT
-rw-r--r--  1 x x   16 Jun 18 21:32 CURRENT.bak
-rw-r--r--  1 x x    0 Jun 18 21:31 LOCK
-rw-r--r--  1 x x 2264 Jun 18 21:32 LOG
-rw-r--r--  1 x x   87 Jun 18 21:32 MANIFEST-000008
```

Reply:
it is not _expected_ to find `~/.sigstore/` there - this is more to _ensure_ that there isn't one

docs suggest removing the whole dir and so that's probably a good idea:
https://docs.sigstore.dev/system_config/public_deployment/#usage